### PR TITLE
ProvidesWithAdvice feature

### PR DESCRIPTION
**Pull request by elandau**
Modification of Tom's ProvidesBuilder to use @Named and maintain access to the InjectionPoint so we can implement advanced features such as conditionals in the future.

### CODE REVIEW

**Comment by elandau** [comment]
Based on this PR https://github.com/Netflix/governator/pull/318

File: governator-providers/build.gradle [hunk 1]
```diff
@@ -0,0 +1,15 @@
+apply plugin: 'java'
+
+dependencies {
+    compile project(':governator-core')
+
+    compile "com.google.inject.extensions:guice-servlet:${guice_version}"
+
+    compile 'com.sun.jersey:jersey-server:1.19'
```
Comment (twicksell) [inline governator-providers/build.gradle]: No need for jersey

File: governator-providers/src/main/java/com/netflix/governator/providers/AdvisableAnnotatedMethodScanner.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        
+        public AdvisedProvider(String name, Annotation annotation, Provider<T> delegate) {
+            this.name = name;
+            this.delegate = delegate;
+        }
+        
+        @Override
+        public T get() {
+            return adviceBindings
+                .stream()
```
Comment (twicksell) [inline governator-providers/src/main/java/com/netflix/governator/providers/AdvisableAnnotatedMethodScanner.java]: So we're totally java 8 now?

File: governator-providers/src/main/java/com/netflix/governator/providers/AdvisableAnnotatedMethodScanner.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        }
+    };
+    
+    static class ProvisionAdviceHolder<T>  {
+        private Binding<T> binding;
+        int order;
+        
+        public ProvisionAdviceHolder(Binding<T> binding, AdviceElementImpl element) {
+            Method member = (Method)element.injectionPoint.getMember();
+            this.order = Optional.ofNullable(member.getAnnotation(Order.class)).map(order -> order.value()).orElse(1000);
```
Comment (twicksell) [inline governator-providers/src/main/java/com/netflix/governator/providers/AdvisableAnnotatedMethodScanner.java]: Why 1000?

**Pull request merged**